include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,367 @@ enum E_DESCRIPTOR_TYPE : uint8_t
+		const core::smart_refctd_ptr<sampler_type>* samplers;
+	};
+
+	// Maps a binding to a local (to descriptor set layout) offset.
+	class CBindingRedirect
+	{
+	public:
 		static constexpr inline uint32_t Invalid = ~0u;
 
-		SBindingRedirect() : bindings(nullptr), offsets(nullptr), count(0ull) {}
+		struct binding_number_t
+		{
+			inline binding_number_t(const uint32_t d) : data(d) {}
+			uint32_t data;
+		};
+
+		struct storage_offset_t
+		{
+			inline storage_offset_t(const uint32_t d) : data(d) {}
+			uint32_t data;
+		};
+
+		inline uint32_t getBindingCount() const { return m_count; }
+
+		// Returns index into the binding property arrays below (including `m_storageOffsets`), for the given binding number `binding`.
+		// Assumes `m_bindingNumbers` is sorted and that there are no duplicate values in it.
+		inline uint32_t searchForBinding(const binding_number_t binding) const
+		{
+			if (!m_bindingNumbers)
+				return Invalid;
+
+			assert(m_storageOffsets && (m_count != 0u));
+
+			auto found = std::lower_bound(m_bindingNumbers, m_bindingNumbers + m_count, binding, [](binding_number_t a, binding_number_t b) -> bool {return a.data < b.data; });
+
+			if ((found >= m_bindingNumbers + m_count) || (found->data != binding.data))
+				return Invalid;
+
+			const uint32_t foundIndex = found - m_bindingNumbers;
+			assert(foundIndex < m_count);
+			return foundIndex;
+		}
+
+		inline binding_number_t getBindingNumber(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return m_bindingNumbers[index];
+		}
+
+		inline core::bitflag<IShader::E_SHADER_STAGE> getStageFlags(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return m_stageFlags[index];
+		}
 
-		SBindingRedirect(const size_t _count) : count(_count)
+		inline uint32_t getCount(const uint32_t index) const
 		{
-			bindings = std::make_unique<uint32_t[]>(count << 1);
-			offsets = bindings.get() + count;
+			assert(index < m_count);
+			return (index == 0u) ? m_storageOffsets[index].data : m_storageOffsets[index].data - m_storageOffsets[index - 1].data;
 		}
 
-		// For a given binding number `binding` in `bindings`, return its offset in `offsets`.
-		inline uint32_t operator[](const uint32_t binding) const
+		inline storage_offset_t getStorageOffset(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return (index == 0u) ? 0u : m_storageOffsets[index - 1];
+		}
+
+		// The follwoing are merely convenience functions for one off use.
+		// If you already have an index (the result of `searchForBinding`) lying around use the above functions for quick lookups, and to avoid unnecessary binary searches.
+
+		inline core::bitflag<IShader::E_SHADER_STAGE> getStageFlags(const binding_number_t binding) const
 		{
-			const uint32_t index = searchForBinding(binding);
+			const auto index = searchForBinding(binding);
 			if (index == Invalid)
 				return Invalid;
```

Comment:
returning `Invalid` as the bitflag? shouldn't it be NONE or something for the bitflag?

Reply:
Yeah you're right, its copypasta.

Reply:
Resolved.